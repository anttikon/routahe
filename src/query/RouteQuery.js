import GraphQlQuery from './GraphQlQuery'
import {get, extend} from 'lodash'

class RouteQuery extends GraphQlQuery {
  constructor() {
    super({
      url: 'https://api.digitransit.fi/routing/v1/routers/hsl/index/graphql',
      command: 'plan',
      schema: {
        itineraries: {
          fields: {
            duration: {},
            legs: {
              fields: {
                startTime: {},
                endTime: {},
                duration: {},
                distance: {},
                mode: {},
                from: {fields: {name: {}}},
                to: {fields: {name: {}}},
                route: {fields: {shortName: {}}}
              }
            }
          }
        }
      }
    })
  }

  preProcess({from, to, date, arriveBy}) {
    const queryJson = {
      from: {lat: from.lat, lon: from.lon},
      to: {lat: to.lat, lon: to.lon},
      numItineraries: 3
    }

    if (date) {
      extend(queryJson, {
        date: date.format('YYYY-MM-DD'),
        time: date.format('HH:mm'),
        arriveBy
      })
    }
    return queryJson
  }

  postProcess(result) {
    return get(result, 'data.plan.itineraries')
  }
}

export default RouteQuery