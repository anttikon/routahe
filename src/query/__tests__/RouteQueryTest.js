import moment from 'moment'
import RouteQuery from '../RouteQuery'

describe('HSL-API', function() {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = 50000

  it('should be able to get route', async() => {
    const date = moment(1481728405559)
    const routes = await new RouteQuery().fetch({
      from: {label: 'Kamppi, Helsinki', lon: 24.931861, lat: 60.169067}, to: {
        label: 'Kannistontie 2, Vantaa',
        lon: 24.841085,
        lat: 60.322768
      }, date, arriveBy: false
    })
    expect(routes).toMatchSnapshot()
  })

  it('should be able to get route with arriveBy', async() => {
    const date = moment(1481728405559)
    const routes = await new RouteQuery().fetch({
      from: {label: 'Kamppi, Helsinki', lon: 24.931861, lat: 60.169067}, to: {
        label: 'Kannistontie 2, Vantaa',
        lon: 24.841085,
        lat: 60.322768
      }, date, arriveBy: true
    })
    expect(routes).toMatchSnapshot()
  })
})