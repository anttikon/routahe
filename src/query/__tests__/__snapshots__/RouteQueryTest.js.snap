exports[`HSL-API should be able to get route 1`] = `
Array [
  Object {
    "duration": 3117,
    "legs": Array [
      Object {
        "distance": 743.0129999999998,
        "duration": 646,
        "endTime": 1481729459000,
        "from": Object {
          "name": "Origin",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481728813000,
        "to": Object {
          "name": "Helsinki",
        },
      },
      Object {
        "distance": 17312.512924842842,
        "duration": 1440,
        "endTime": 1481730900000,
        "from": Object {
          "name": "Helsinki",
        },
        "mode": "RAIL",
        "route": Object {
          "shortName": "P",
        },
        "startTime": 1481729460000,
        "to": Object {
          "name": "Kivistö",
        },
      },
      Object {
        "distance": 1258.6750000000002,
        "duration": 1029,
        "endTime": 1481731930000,
        "from": Object {
          "name": "Kivistö",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481730901000,
        "to": Object {
          "name": "Destination",
        },
      },
    ],
  },
  Object {
    "duration": 3343,
    "legs": Array [
      Object {
        "distance": 468.127,
        "duration": 383,
        "endTime": 1481729399000,
        "from": Object {
          "name": "Origin",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481729016000,
        "to": Object {
          "name": "Hki linja-autoasema",
        },
      },
      Object {
        "distance": 17452.41190604001,
        "duration": 1680,
        "endTime": 1481731080000,
        "from": Object {
          "name": "Hki linja-autoasema",
        },
        "mode": "BUS",
        "route": Object {
          "shortName": "455",
        },
        "startTime": 1481729400000,
        "to": Object {
          "name": "Keimola",
        },
      },
      Object {
        "distance": 1563.4159999999997,
        "duration": 1278,
        "endTime": 1481732359000,
        "from": Object {
          "name": "Keimola",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481731081000,
        "to": Object {
          "name": "Destination",
        },
      },
    ],
  },
  Object {
    "duration": 3117,
    "legs": Array [
      Object {
        "distance": 743.0129999999998,
        "duration": 646,
        "endTime": 1481730059000,
        "from": Object {
          "name": "Origin",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481729413000,
        "to": Object {
          "name": "Helsinki",
        },
      },
      Object {
        "distance": 17312.512924842842,
        "duration": 1440,
        "endTime": 1481731500000,
        "from": Object {
          "name": "Helsinki",
        },
        "mode": "RAIL",
        "route": Object {
          "shortName": "P",
        },
        "startTime": 1481730060000,
        "to": Object {
          "name": "Kivistö",
        },
      },
      Object {
        "distance": 1258.6750000000002,
        "duration": 1029,
        "endTime": 1481732530000,
        "from": Object {
          "name": "Kivistö",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481731501000,
        "to": Object {
          "name": "Destination",
        },
      },
    ],
  },
]
`;

exports[`HSL-API should be able to get route with arriveBy 1`] = `
Array [
  Object {
    "duration": 3117,
    "legs": Array [
      Object {
        "distance": 743.0129999999998,
        "duration": 646,
        "endTime": 1481725859000,
        "from": Object {
          "name": "Origin",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481725213000,
        "to": Object {
          "name": "Helsinki",
        },
      },
      Object {
        "distance": 17312.512924842842,
        "duration": 1440,
        "endTime": 1481727300000,
        "from": Object {
          "name": "Helsinki",
        },
        "mode": "RAIL",
        "route": Object {
          "shortName": "P",
        },
        "startTime": 1481725860000,
        "to": Object {
          "name": "Kivistö",
        },
      },
      Object {
        "distance": 1257.7230000000002,
        "duration": 1029,
        "endTime": 1481728330000,
        "from": Object {
          "name": "Kivistö",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481727301000,
        "to": Object {
          "name": "Destination",
        },
      },
    ],
  },
  Object {
    "duration": 3117,
    "legs": Array [
      Object {
        "distance": 743.0129999999998,
        "duration": 646,
        "endTime": 1481725259000,
        "from": Object {
          "name": "Origin",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481724613000,
        "to": Object {
          "name": "Helsinki",
        },
      },
      Object {
        "distance": 17312.512924842842,
        "duration": 1440,
        "endTime": 1481726700000,
        "from": Object {
          "name": "Helsinki",
        },
        "mode": "RAIL",
        "route": Object {
          "shortName": "P",
        },
        "startTime": 1481725260000,
        "to": Object {
          "name": "Kivistö",
        },
      },
      Object {
        "distance": 1257.7230000000002,
        "duration": 1029,
        "endTime": 1481727730000,
        "from": Object {
          "name": "Kivistö",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481726701000,
        "to": Object {
          "name": "Destination",
        },
      },
    ],
  },
  Object {
    "duration": 3834,
    "legs": Array [
      Object {
        "distance": 897.4499999999999,
        "duration": 765,
        "endTime": 1481725259000,
        "from": Object {
          "name": "Origin",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481724494000,
        "to": Object {
          "name": "Helsinki",
        },
      },
      Object {
        "distance": 30410.431214842498,
        "duration": 2040,
        "endTime": 1481727300000,
        "from": Object {
          "name": "Helsinki",
        },
        "mode": "RAIL",
        "route": Object {
          "shortName": "I",
        },
        "startTime": 1481725260000,
        "to": Object {
          "name": "Kivistö",
        },
      },
      Object {
        "distance": 1256.23,
        "duration": 1027,
        "endTime": 1481728328000,
        "from": Object {
          "name": "Kivistö",
        },
        "mode": "WALK",
        "route": null,
        "startTime": 1481727301000,
        "to": Object {
          "name": "Destination",
        },
      },
    ],
  },
]
`;
